### PR TITLE
Apply state diffs

### LINKED ISSUE #90: feat: Apply State Diff ✨
# Feature Request

**Describe the Feature Request**

We should compute the state diff after `invoke` and `deploy` transactions.
The state trie should also be updated after `declare` and `deploy` transactions.

**Describe Preferred Solution**

- This involves writing wrappers for some `starknet-api` structs.
- Then the `StorageView` mapping in the `starknet` pallet should be adequately updated.

**Related Code**

https://github.com/starkware-libs/starknet-api/blob/main/src/state.rs#L21

**Additional Context**

- https://docs.starknet.io/documentation/architecture_and_concepts/Data_Availability/on-chain-data/#v0.11.0_format
- https://david-barreto.com/starknets-architecture-review/

**If the feature request is approved, would you be willing to submit a PR?**
_(Help can be provided if you need assistance submitting a PR)_

- [ ] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by LucasLvy**
# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [x] Documentation content changes
- [x] Testing
- [ ] Other (please describe):

# What is the current behavior?

Issue Number: Fixes #90 

# What is the new behavior?

- Apply state diffs to substrate storage

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

# Other information

### CODE REVIEW

File: crates/pallets/starknet/src/lib.rs [hunk 1]
```diff
@@ -654,6 +654,29 @@ pub mod pallet {
             Ok(())
         }
 
+        pub fn apply_state_diffs(state: &CachedState<DictStateReader>) -> Result<(), StateDiffError> {
```
Comment (abdelhamidbakhta) [inline crates/pallets/starknet/src/lib.rs:657]: please add rust doc

Comment (abdelhamidbakhta) [reply crates/pallets/starknet/src/lib.rs:657]: please add comments inside the function explaining the different steps

**Review by abdelhamidbakhta — CHANGES_REQUESTED** [review]
left some comments

File: crates/primitives/starknet/src/transaction/types.rs [hunk 2]
```diff
@@ -77,3 +77,13 @@ pub enum EventError {
     /// Provided from address is invalid.
     InvalidFromAddress,
 }
+
+/// Error enum wrapper for events.
```
Comment (EvolveArt) [inline crates/primitives/starknet/src/transaction/types.rs:81]: wrong comment

**Review by EvolveArt — CHANGES_REQUESTED** [review]

File: crates/pallets/starknet/src/tests.rs [hunk 3]
```diff
@@ -443,3 +443,55 @@ fn given_hardcoded_contract_run_declare_none_then_it_fails() {
...
+                    storage_var_selector,
+                    storage_var_val,
+                ],
+                contract_address_bytes,
+                contract_address_bytes,
+            ),
+            None,
+        );
+
+        // Cannot declare a class with None
```
[the remark this answers is not in the document]
Comment (LucasLvy) [reply crates/pallets/starknet/src/tests.rs:493]: You mean explain what a selector is ?

File: crates/primitives/starknet/src/execution/mod.rs [hunk 4]
```diff
@@ -63,6 +63,16 @@ impl From<ContractClass> for ContractClassWrapper {
         }
     }
 }
+impl From<&starknet_api::state::ContractClass> for ContractClassWrapper {
```
[the remark this answers is not in the document]
Comment (LucasLvy) [reply crates/primitives/starknet/src/execution/mod.rs:66]: it was to stay coherent with the one above but sure can switch

File: crates/pallets/starknet/src/tests.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (EvolveArt) [reply crates/pallets/starknet/src/tests.rs:493]: No just what function name does it refer to

File: crates/primitives/starknet/src/execution/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (EvolveArt) [reply crates/primitives/starknet/src/execution/mod.rs:66]: Alright well if you could switch both would be awesome

[the remark this answers is not in the document]
Comment (LucasLvy) [reply crates/primitives/starknet/src/execution/mod.rs:66]: wasn't used so removed it

**Review by EvolveArt — APPROVED** [review]

**Review by abdelhamidbakhta — APPROVED** [review]
lgtm

**Pull request merged**